namecache/namecache.go

```diff
@@ -0,0 +1,258 @@
+package namecache
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	bstore "github.com/ipfs/go-ipfs/blocks/blockstore"
+	merkledag "github.com/ipfs/go-ipfs/merkledag"
+	namesys "github.com/ipfs/go-ipfs/namesys"
+	path "github.com/ipfs/go-ipfs/path"
+	pin "github.com/ipfs/go-ipfs/pin"
+	uio "github.com/ipfs/go-ipfs/unixfs/io"
+
+	cid "gx/ipfs/QmNp85zy9RLrQ5oQD4hPyS39ezrrXpcaa7R4Y9kxdWQLLQ/go-cid"
+	node "gx/ipfs/QmPN7cwmpcc4DWXb4KTB9dNAJgjuPY69h3npsMfhRrQL9c/go-ipld-format"
+	logging "gx/ipfs/QmSpJByNKFX1sCsHBEp3R73FL4NF6FnQTEGyNAXHm2GS52/go-log"
+)
+
+const (
+	DefaultFollowInterval = 1 * time.Hour
+	resolveTimeout        = 1 * time.Minute
+)
+
+var log = logging.Logger("namecache")
+
+// NameCache represents a following cache of names
+type NameCache interface {
+	// Follow starts following name, pinning it if pinit is true
+	Follow(name string, pinit bool, followInterval time.Duration) error
+	// Unofollow cancels a follow
+	Unfollow(name string) error
+	// ListFollows returns a list of followed names
+	ListFollows() []string
+}
+
+type nameCache struct {
+	nsys    namesys.NameSystem
+	pinning pin.Pinner
+	dag     merkledag.DAGService
+	bstore  bstore.GCBlockstore
+
+	ctx     context.Context
+	follows map[string]func()
+	mx      sync.Mutex
+}
+
+func NewNameCache(ctx context.Context, nsys namesys.NameSystem, pinning pin.Pinner, dag merkledag.DAGService, bstore bstore.GCBlockstore) NameCache {
+	return &nameCache{
+		ctx:     ctx,
+		nsys:    nsys,
+		pinning: pinning,
+		dag:     dag,
+		bstore:  bstore,
+		follows: make(map[string]func()),
+	}
+}
+
+// Follow spawns a goroutine that periodically resolves a name
+// and (when pinit is true) pins it in the background
+func (nc *nameCache) Follow(name string, pinit bool, followInterval time.Duration) error {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	if !strings.HasPrefix(name, "/ipns/") {
+		name = "/ipns/" + name
+	}
+
+	if _, ok := nc.follows[name]; ok {
+		return fmt.Errorf("Already following %s", name)
+	}
+
+	ctx, cancel := context.WithCancel(nc.ctx)
+	go nc.followName(ctx, name, pinit, followInterval)
+	nc.follows[name] = cancel
+
+	return nil
+}
+
+// Unfollow cancels a follow
+func (nc *nameCache) Unfollow(name string) error {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	if !strings.HasPrefix(name, "/ipns/") {
+		name = "/ipns/" + name
+	}
+
+	cancel, ok := nc.follows[name]
+	if !ok {
+		return fmt.Errorf("Unknown name %s", name)
+	}
+
+	cancel()
+	delete(nc.follows, name)
+	return nil
+}
+
+// ListFollows returns a list of names currently being followed
+func (nc *nameCache) ListFollows() []string {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	follows := make([]string, 0, len(nc.follows))
+	for name, _ := range nc.follows {
+		follows = append(follows, name)
+	}
+
+	return follows
+}
+
+func (nc *nameCache) followName(ctx context.Context, name string, pinit bool, followInterval time.Duration) {
```

Comment:
I'd use `dopin` instead of `pinit` as it's already used in few places.

Reply:
sure enough!